rflx/model.py

```diff
@@ -549,18 +570,35 @@ def __check_length_expression(self, link: Link) -> None:
                 )
 
     def __type_constraints(self, expr: Expr) -> Sequence[Expr]:
+        def get_constraints(aggregate: Aggregate, field: Variable) -> Sequence[Expr]:
+            comp = self.types[Field(field.name)]
+            assert isinstance(comp, Composite)
+            result = Equal(Mul(aggregate.length, comp.element_size), Length(field))
+            if isinstance(comp, Array) and isinstance(comp.element_type, Scalar):
+                return [
+                    result,
+                    *comp.element_type.constraints(name=comp.element_type.name, proof=True),
+                ]
+            return [result]
+
         literals = qualified_literals(self.types, self.package)
         scalar_types = [
-            (n, t)
-            for n, t in [
-                (v.name, self.types[Field(v.identifier)])
-                for v in expr.variables()
-                if v.name not in [*literals, "Message", "Final"]
-            ]
-            if isinstance(t, Scalar)
+            (f.name, t)
+            for f, t in self.types.items()
+            if isinstance(t, Scalar) and f.name not in [*literals, "Message", "Final"]
         ]
 
-        return [c for n, t in scalar_types for c in t.constraints(name=n, proof=True)]
+        aggregate_constraints: List[Expr] = []
+        for r in expr.findall(lambda x: isinstance(x, (Equal, NotEqual))):
+            if isinstance(r, (Equal, NotEqual)):
```

Comment:
An assert would make more sense here (and will also improve code coverage).

Reply:
Done in c157478